scripts/generate_header.py

```diff
@@ -169,43 +177,97 @@ def generate_offsets(device_name: str, reg_list: t.List[Register]) -> str:
     for a_reg in reg_list:
-        name = a_reg.name.upper()
+        if a_reg.name == 'reserved':
+            continue
+        name = a_reg.name.upper().strip().replace(" ", "")
+        group = a_reg.group.upper().strip().replace(" ", "")
         offset = a_reg.offset
-        macro_line = f'#define METAL_{cap_device}_{name} {offset}'
+        width = a_reg.width
+
+        if name.startswith(group.split('_')[-1] + "_"):
+            group = "_".join(group.split('_')[:-1])
+
+        if group is not "":
+            macro_line = f'#define METAL_{cap_device}_{group}_{name} {offset}\n'
+            NAME_DICT[f'METAL_{cap_device}_{group}_{name}'] += 1
+            macro_line += f'#define METAL_{cap_device}_{group}_{name}_BYTE {offset >> 3}\n'
+            macro_line += f'#define METAL_{cap_device}_{group}_{name}_BIT {offset & 0x7}\n'
+            macro_line += f'#define METAL_{cap_device}_{group}_{name}_WIDTH {width}\n'
+        else:
+            macro_line = f'#define METAL_{cap_device}_{name} {offset}\n'
+            NAME_DICT[f'METAL_{cap_device}_{name}'] += 1
+            macro_line += f'#define METAL_{cap_device}_{name}_BYTE {offset >> 3}\n'
+            macro_line += f'#define METAL_{cap_device}_{name}_BIT {offset & 0x7}\n'
+            macro_line += f'#define METAL_{cap_device}_{name}_WIDTH {width}\n'
+
         rv.append(macro_line)
 
     return '\n'.join(rv)
 
 
-def generate_widths(device_name: str, reg_list: t.List[Register]) -> str:
-    rv: t.List[str] = []
+@dataclass(frozen=True)
+class Interrupt:
+    number: int
+    name: str
 
-    cap_device = device_name.upper()
-    for a_reg in reg_list:
-        name = a_reg.name.upper()
-        width = a_reg.width
-        macro_line = f'#define METAL_{cap_device}_{name}_WIDTH {width}'
-        rv.append(macro_line)
+    @staticmethod
+    def make_interrupt(number, name=''):
+        return Interrupt(number, name)
+
+
+def generate_interrupt_list(object_model: JSONType) -> t.List[Interrupt]:
+    p = walk(object_model)
+    p = filter(lambda x: '_types' in x, p)
+    p = filter(lambda x: f'OMInterrupt' in x['_types'], p)
+
+    rv = []
+    for an_interrupt in p:
+        number = an_interrupt['numberAtReceiver']
+        name = an_interrupt.get('name', '')
+        if '@' in name:
+            name = ''
+        rv.append(Interrupt.make_interrupt(number, name))
+
+    return rv
+
+
+def generate_interrupt_defines(int_list: t.List[Interrupt], device: str) -> str:
+    rv = []
+    interrupts = []
+    dev = device.upper().replace(' ', '')
+
+    for an_interrupt in int_list:
+        number = an_interrupt.number
+        if an_interrupt.name:
+            name = an_interrupt.name.upper().replace(' ', '')
+            rv.append(f'#define {dev}_{name}_IT {number}')
+        interrupts.append(an_interrupt.number)
+
+    rv.append(f'#define {dev}_INTERRUPT_COUNT {len(interrupts)}')
+    rv.append(f'#define {dev}_INTERRUPTS {{ {",".join(map(str,interrupts)) } }}')
```

Comment:
I'm not sure if this will work well if there are multiple instances of a device in a design. Although it might be safe in most cases to assume that the relative offsets of the interrupt numbers are consistent across different device instances, they will definitely have different absolute numbers.

We probably need a data structure here that describes at least the base of the block of interrupt numbers that is assigned to the device.